radix-engine-utils/src/rocks_db_metrics/mod.rs

```diff
@@ -0,0 +1,544 @@
+        export_graph_and_print_summary(
+            &mut substate_db,
+            "InMemoryDB random reads",
+            &inmem_data,
+            &inmem_output_data,
+            "/tmp/scrypto_inmem_1.png",
+            "95th percentile of reads",
+        )
+        .unwrap();
+
+        // Calculate RocksDB - InMemory diff and export results
+        export_graph_and_print_summary_for_two_series(
+            "RocksDB - InMemoryDB random reads",
+            &rocksdb_output_data,
+            &inmem_output_data,
+            "/tmp/scrypto_diff_1.png",
+        )
+        .unwrap();
+    }
+
+    fn drop_edge_values<S: SubstateDatabase + CommittableSubstateDatabase>(
+        substate_store: &mut SubstateStoreWithMetrics<S>,
+    ) {
+        for (_, v) in substate_store.read_metrics.borrow_mut().iter_mut() {
+            v.sort();
+            v.pop();
+            v.remove(0);
+        }
+    }
+
+    pub fn calculate_percent_to_max_points<S: SubstateDatabase + CommittableSubstateDatabase>(
+        substate_store: &mut SubstateStoreWithMetrics<S>,
+        percent: f32,
+    ) -> Vec<(f32, f32)> {
+        assert!(percent <= 100f32);
+        let mut output_values = Vec::new();
+        let mut binding = substate_store.read_metrics.borrow_mut();
+        for (k, v) in binding.iter_mut() {
+            v.sort();
+            let idx = (((v.len() - 1) as f32 * percent) / 100f32).round() as usize;
+            output_values.push((*k as f32, v[idx].as_micros() as f32));
+        }
+        output_values
+    }
+
+    fn prepare_db<S: SubstateDatabase + CommittableSubstateDatabase>(
+        substate_db: &mut S,
+    ) -> Vec<(DbPartitionKey, DbSortKey, usize)> {
+        let mut data_index_vector: Vec<(DbPartitionKey, DbSortKey, usize)> =
+            Vec::with_capacity(MAX_SIZE);
+
+        println!("Preparing database...");
+        let mut p_key_cnt = 1u32;
+        //let mut substate_db = SubstateStoreWithMetrics::new_rocksdb(path.clone());
+        let mut sort_key_value: usize = 0;
+        for size in (MIN_SIZE..=MAX_SIZE).step_by(SIZE_STEP) {
+            let mut input_data = DatabaseUpdates::new();
+            for _ in 0..COUNT {
+                let value = DatabaseUpdate::Set(vec![1; size]);
+
+                let plain_bytes = sort_key_value.to_be_bytes().to_vec();
+                let mut hashed_prefix: Vec<u8> =
+                    Blake2b::<U32>::digest(plain_bytes.clone()).to_vec();
+                hashed_prefix.extend(plain_bytes);
+
+                let sort_key = DbSortKey(hashed_prefix);
+                sort_key_value += 1;
+
+                let mut partition = PartitionUpdates::new();
+                partition.insert(sort_key.clone(), value);
+
+                let partition_key = DbPartitionKey(p_key_cnt.to_be_bytes().to_vec());
+                p_key_cnt += 1;
+
+                data_index_vector.push((partition_key.clone(), sort_key, size));
+
+                input_data.insert(partition_key, partition);
```

Comment:
We may use `to_db_partition_key(node_id: &NodeId, partition_num: PartitionNumber)` and `to_db_sort_key(key: &SubstateKey)` to generate data, as this is how RE uses substate store.

Also, we will need to re-run the tests by using the interface Jakub will provide, as the default store doesn't use merkle tree (which has huge overhead).

Reply:
I refactored db preparation code to use `to_db_partition_key()` and `to_db_sort_key()`, but this requires moving `db_key_mapper.rs` file from `radix-engine` project because of circular dependency (`radix-engine` uses `radix-engine-profiling`). I moved it to `radix-engine-store-interface` project.